diff --git a/frame/base/bli_gks.c b/frame/base/bli_gks.c
index 8904a9e6..65427f66 100644
--- a/frame/base/bli_gks.c
+++ b/frame/base/bli_gks.c
@@ -351,6 +351,7 @@ bool bli_gks_cntx_ukr_is_ref
      )
 {
 	cntx_t ref_cntx;
+  memset( &ref_cntx, 0, sizeof( ref_cntx ) );
 
 	// Initialize a context with reference kernels for the arch_t id queried
 	// via bli_arch_query_id().
@@ -376,6 +377,7 @@ bool bli_gks_cntx_ukr2_is_ref
      )
 {
 	cntx_t ref_cntx;
+  memset( &ref_cntx, 0, sizeof( ref_cntx ) );
 
 	// Initialize a context with reference kernels for the arch_t id queried
 	// via bli_arch_query_id().
